File: agents.py
"""
Multi-Agent Network Engine for Aegis
5 Autonomous AI Agents: CEO, Strategy, Marketing, Sales, Finance
"""

import random
import time
from datetime import datetime


# ─────────────────────────────────────────────
# AGENT DEFINITIONS
# ─────────────────────────────────────────────

class CEO_Agent:
    name = "CEO Agent"
    icon = "👔"
    color = "#3b82f6"
    bg = "rgba(59,130,246,0.15)"
    role = "Chief Intelligence Officer"
    responsibility = "Synthesizes all agent outputs into executive-level insights, sets strategic priorities, and communicates board-ready summaries."
    input_desc = "Business onboarding data (company profile, industry, revenue, goals)"
    output_desc = "Executive Summary, Board Briefing, Business Health Score, Strategic Mandate"

    @staticmethod
    def run(business_data: dict) -> dict:
        industry = business_data.get("industry", "Technology")
        company = business_data.get("company_name", "Your Company")
        revenue = business_data.get("annual_revenue", 1_000_000)
        audience = business_data.get("target_audience", "SMBs")

        industry_context = {
            "Technology": ("SaaS expansion", "developer-led growth", "ARR acceleration"),
            "Healthcare": ("patient acquisition", "compliance-first growth", "value-based care"),
            "Retail / E-Commerce": ("omnichannel dominance", "AOV optimization", "retention flywheel"),
            "Financial Services": ("AUM growth", "regulatory arbitrage", "fintech disruption"),
            "Education": ("learner acquisition", "outcome-based positioning", "cohort monetization"),
            "Manufacturing": ("operational efficiency", "supply chain resilience", "B2B pipeline"),
            "Real Estate": ("deal velocity", "portfolio diversification", "proptech integration"),
            "Consulting / Professional Services": ("retainer growth", "thought leadership", "IP monetization"),
        }.get(industry, ("market expansion", "customer acquisition", "revenue diversification"))

        summary = (
            f"{company} is positioned at a critical inflection point within the {industry} sector. "
            f"With an annual revenue baseline of ${revenue:,.0f}, the immediate mandate is to accelerate "
            f"{industry_context[0]} through a disciplined {industry_context[1]} approach, "
            f"targeting {audience} as the primary growth vector. "
            f"Our multi-agent analysis has identified three strategic horizons: "
            f"(1) 0–90 day quick wins focused on conversion optimization and pipeline velocity, "
            f"(2) 90–180 day structural investments in brand authority and {industry_context[2]}, "
            f"and (3) 12-month market dominance through category leadership. "
            f"The Aegis Growth OS has computed a composite Business Health Score reflecting "
            f"market readiness, operational capacity, and competitive positioning."
        )

        health_score = random.randint(62, 91)
        growth_score = random.randint(55, 88)
        return {
            "summary": summary,
            "health_score": health_score,
            "growth_score": growth_score,
            "mandate": f"Achieve {industry_context[2]} within 12 months via {industry_context[1]}",
            "timestamp": datetime.now().strftime("%H:%M:%S"),
        }


class Strategy_Agent:
    name = "Strategy Agent"
    icon = "🧭"
    color = "#8b5cf6"
    bg = "rgba(139,92,246,0.15)"
    role = "Chief Strategy Architect"
    responsibility = "Constructs comprehensive growth strategies using market intelligence, competitive analysis, and 5D framework alignment."
    input_desc = "CEO mandate, business profile, market benchmarks, industry vertical data"
    output_desc = "Growth Strategy Document, SWOT Analysis, Competitive Positioning, KPI Targets"

    @staticmethod
    def run(business_data: dict, ceo_output: dict) -> dict:
        industry = business_data.get("industry", "Technology")
        company = business_data.get("company_name", "Your Company")
        revenue = business_data.get("annual_revenue", 1_000_000)

        strategies = {
            "Technology": {
                "primary": "Product-Led Growth (PLG) with enterprise upsell motion",
                "pillars": ["Freemium to paid conversion funnel", "Developer community building", "API-first integrations", "Usage-based pricing tiers"],
                "kpis": {"MRR Growth": "15% MoM", "CAC Payback": "< 9 months", "NRR": "> 120%", "Trial-to-Paid": "> 25%"},
                "markets": ["Mid-market SaaS", "Enterprise IT", "Startup ecosystem"],
                "competitive_moat": "Deep API ecosystem + switching cost via data gravity",
            },
            "Healthcare": {
                "primary": "Trust-Led Acquisition with referral network amplification",
                "pillars": ["HIPAA-compliant digital touchpoints", "Physician referral programs", "Outcome-based case studies", "Telehealth channel expansion"],
                "kpis": {"Patient Acquisition Cost": "< $120", "Retention Rate": "> 78%", "NPS": "> 55", "Digital Appointments": "> 40%"},
                "markets": ["Urban outpatient clinics", "Insurance networks", "Corporate wellness"],
                "competitive_moat": "Regulatory compliance expertise + established trust scores",
            },
            "Retail / E-Commerce": {
                "primary": "Retention-First Flywheel with DTC channel dominance",
                "pillars": ["Loyalty program with tiered rewards", "Personalization engine", "Social commerce expansion", "Subscription bundles"],
                "kpis": {"ROAS": "> 4.5x", "LTV:CAC": "> 3:1", "Cart Abandonment": "< 62%", "Repeat Purchase Rate": "> 35%"},
                "markets": ["Millennial shoppers 25–40", "Mobile-first buyers", "Sustainability-conscious consumers"],
                "competitive_moat": "Brand community + first-party data advantage",
            },
        }.get(industry, {
            "primary": "Market Penetration with authority-led positioning",
            "pillars": ["Content moat construction", "Strategic partnership matrix", "Direct sales acceleration", "Digital channel optimization"],
            "kpis": {"Revenue Growth": "25% YoY", "Market Share": "+3–5%", "Lead Velocity Rate": "+20% MoM", "CAC Reduction": "15%"},
            "markets": ["Primary TAM segments", "Adjacent verticals", "International expansion"],
            "competitive_moat": "Domain expertise + network effects",
        })

        growth_projection = {
            "Q1": revenue * 1.08,
            "Q2": revenue * 1.18,
            "Q3": revenue * 1.31,
            "Q4": revenue * 1.47,
        }

        return {
            "strategy": strategies,
            "growth_projection": growth_projection,
            "mandate": ceo_output.get("mandate", ""),
            "timestamp": datetime.now().strftime("%H:%M:%S"),
        }


class Marketing_Agent:
    name = "Marketing Agent"
    icon = "📣"
    color = "#06b6d4"
    bg = "rgba(6,182,212,0.15)"
    role = "Chief Growth Marketer"
    responsibility = "Generates multi-channel campaign strategies, ad copy, email funnels, and brand positioning assets from the strategy blueprint."
    input_desc = "Growth strategy, target audience profile, budget parameters, channel preferences"
    output_desc = "Ad Campaigns, Email Funnels, Channel Mix, Content Calendar, Brand Voice Guidelines"

    @staticmethod
    def run(business_data: dict, strategy_output: dict) -> dict:
        industry = business_data.get("industry", "Technology")
        audience = business_data.get("target_audience", "SMBs")
        company = business_data.get("company_name", "Your Company")

        campaigns = {
            "Technology": {
                "hero_ad": f"Stop losing deals to competitors with smarter tools. {company} gives your team the AI advantage — measurable ROI in 30 days or your money back.",
                "email_subject": "Your competitors just switched. Here's why you should too →",
                "email_body": f"Hi [First Name],\n\nI noticed your team is scaling fast — and I wanted to reach out at the right time.\n\n{company} has helped 200+ {audience} companies cut their tool spend by 40% while tripling output. \n\nCan I show you a 12-minute demo this week?\n\n[CTA: Book Your Slot]",
                "channels": ["LinkedIn Ads", "Google Search", "Product Hunt", "Dev Communities", "Webinars"],
                "hook": "The AI stack your competitors don't want you to know about.",
            },
            "Healthcare": {
                "hero_ad": f"Your patients deserve faster, smarter care. {company} modernizes your practice without disrupting your workflow — HIPAA compliant, zero downtime.",
                "email_subject": "How [Similar Practice] saw 40% more appointments in 60 days →",
                "email_body": f"Dear Dr. [Last Name],\n\nPatient acquisition is getting harder and more expensive. We get it.\n\n{company} helped practices like yours attract and retain 40% more patients through evidence-based digital outreach — fully compliant.\n\n[CTA: See Case Study]",
                "channels": ["Medical Journals", "LinkedIn", "Healthcare Conferences", "Referral Programs", "Local SEO"],
                "hook": "Modern care starts with a modern practice.",
            },
            "Retail / E-Commerce": {
                "hero_ad": f"Your store. Supercharged. {company} turns browsers into buyers and one-time shoppers into lifelong fans — with AI-powered personalization.",
                "email_subject": "🛒 Left something behind? Here's 15% off to come back →",
                "email_body": f"Hey [First Name],\n\nWe noticed you were checking out [Product] — great choice, by the way.\n\nFor the next 24 hours, we're offering you an exclusive 15% discount. This is just for you.\n\n[CTA: Complete Your Order] ⏰ Expires soon.\n\nP.S. — [Social Proof: 847 people bought this week]",
                "channels": ["Meta/Instagram Ads", "Google Shopping", "TikTok", "Email Automations", "SMS Flows"],
                "hook": "One-click to their new favorite brand.",
            },
        }.get(industry, {
            "hero_ad": f"{company}: The smart choice for ambitious {audience}. Powered by AI. Proven by results.",
            "email_subject": "3 growth levers you're probably not using →",
            "email_body": f"Hi [First Name],\n\n{company} has helped teams like yours unlock hidden revenue opportunities. Here are 3 strategies working right now...\n\n[CTA: Get the Full Playbook]",
            "channels": ["LinkedIn", "Google Ads", "Content Marketing", "Email", "Partnerships"],
            "hook": "Smarter strategy. Faster growth.",
        })

        content_calendar = [
            {"week": "Week 1", "content": "Brand awareness campaign launch", "channel": "LinkedIn + Google"},
            {"week": "Week 2", "content": "Case study + social proof push", "channel": "Email + Retargeting"},
            {"week": "Week 3", "content": "Lead magnet deployment", "channel": "Content + SEO"},
            {"week": "Week 4", "content": "Conversion campaign + limited offer", "channel": "All Channels"},
        ]

        return {
            "campaigns": campaigns,
            "content_calendar": content_calendar,
            "timestamp": datetime.now().strftime("%H:%M:%S"),
        }


class Sales_Agent:
    name = "Sales Agent"
    icon = "🎯"
    color = "#10b981"
    bg = "rgba(16,185,129,0.15)"
    role = "Chief Revenue Officer"
    responsibility = "Builds outbound sales sequences, lead scoring models, pipeline stages, and conversion frameworks to maximize revenue generation."
    input_desc = "Marketing campaigns, ICP definition, competitive positioning, pricing model"
    output_desc = "Sales Scripts, Outbound Sequences, Pipeline Stages, Lead Scoring Model, Objection Handling"

    @staticmethod
    def run(business_data: dict, marketing_output: dict) -> dict:
        industry = business_data.get("industry", "Technology")
        audience = business_data.get("target_audience", "SMBs")
        company = business_data.get("company_name", "Your Company")

        pipeline = [
            {"stage": "Lead Capture", "conversion": "100%", "avg_days": 0, "action": "ICP qualification via AI scoring"},
            {"stage": "Discovery Call", "conversion": "35%", "avg_days": 3, "action": "Needs analysis + pain mapping"},
            {"stage": "Demo / Proposal", "conversion": "60%", "avg_days": 7, "action": "ROI presentation + customization"},
            {"stage": "Negotiation", "conversion": "70%", "avg_days": 14, "action": "Champion-based closing + legal review"},
            {"stage": "Closed Won", "conversion": "65%", "avg_days": 21, "action": "Contract execution + onboarding trigger"},
        ]

        outbound_sequence = [
            {"day": "Day 1", "touch": "Personalized LinkedIn connection + voice note", "goal": "Initial awareness"},
            {"day": "Day 3", "touch": "Email: Industry insight + problem framing", "goal": "Establish credibility"},
            {"day": "Day 7", "touch": "Follow-up: Case study relevant to their vertical", "goal": "Build trust"},
            {"day": "Day 12", "touch": "Phone call + voicemail drop", "goal": "Direct engagement"},
            {"day": "Day 18", "touch": "Video email (Loom) with personalized demo preview", "goal": "Create urgency"},
            {"day": "Day 25", "touch": "Final value-add email: ROI calculator", "goal": "Conversion or recycle"},
        ]

        discovery_script = f"""OPENING:
"Hi [Name], this is [Rep] from {company}. I know your time is valuable — I'll keep this focused. 
We've helped [2–3 similar companies in {industry}] solve [core pain point]. I'd love to understand 
if the same challenge resonates with your team. Do you have 8 minutes?"

DISCOVERY QUESTIONS:
1. "Walk me through how you currently handle [key process]?"
2. "Where does that process break down most often?"
3. "What would perfect look like 12 months from now?"
4. "If we could achieve [outcome], what would that mean for your {audience} targets?"

OBJECTION HANDLING:
• "We already have a solution" → "That's great — what's the one thing you wish it did better?"
• "Not the right time" → "Completely understand — what would make it the right time?"
• "Too expensive" → "Fair — let's quantify the cost of NOT solving this. Can I share a quick calc?"

CLOSE:
"Based on what you've shared, I think we can make a real impact. Can we lock in 30 minutes next week 
for a tailored walkthrough? I'll bring specific metrics from your industry vertical."
"""

        lead_score_criteria = [
            {"factor": "Industry Match", "weight": "25%", "score_range": "0–25"},
            {"factor": "Company Size (employees)", "weight": "20%", "score_range": "0–20"},
            {"factor": "Intent Signal (content downloads)", "weight": "20%", "score_range": "0–20"},
            {"factor": "Engagement Score (email/web)", "weight": "20%", "score_range": "0–20"},
            {"factor": "Decision-Maker Contact", "weight": "15%", "score_range": "0–15"},
        ]

        revenue_opportunity = int(business_data.get("annual_revenue", 1_000_000) * random.uniform(0.18, 0.35))
        lead_score = random.randint(68, 94)

        return {
            "pipeline": pipeline,
            "outbound_sequence": outbound_sequence,
            "discovery_script": discovery_script,
            "lead_score_criteria": lead_score_criteria,
            "revenue_opportunity": revenue_opportunity,
            "lead_score": lead_score,
            "timestamp": datetime.now().strftime("%H:%M:%S"),
        }


class Finance_Agent:
    name = "Finance Agent"
    icon = "💹"
    color = "#f59e0b"
    bg = "rgba(245,158,11,0.15)"
    role = "Chief Financial Intelligence Officer"
    responsibility = "Runs financial risk assessments, models cash flow projections, identifies margin leakage, and flags compliance risks."
    input_desc = "Revenue data, growth projections, cost structure, market benchmarks"
    output_desc = "Risk Alerts, Financial Health Metrics, Cash Flow Projections, Unit Economics Model"

    @staticmethod
    def run(business_data: dict, strategy_output: dict) -> dict:
        revenue = business_data.get("annual_revenue", 1_000_000)
        industry = business_data.get("industry", "Technology")

        # Generate risk alerts
        risk_alerts = []

        risk_templates = {
            "Technology": [
                {"level": "red", "title": "Churn Rate Threshold", "desc": "Simulated churn rate at 8.2% — above SaaS benchmark of 5%. Immediate retention intervention required."},
                {"level": "amber", "title": "CAC Trending Up", "desc": "Customer acquisition cost has increased 22% QoQ. Organic channel investment recommended."},
                {"level": "amber", "title": "Runway Alert", "desc": "At current burn rate, runway is estimated at 11.4 months. Series A timeline should be accelerated."},
                {"level": "green", "title": "Gross Margin Healthy", "desc": "Gross margin at 74% — above industry median. Strong foundation for R&D reinvestment."},
            ],
            "Healthcare": [
                {"level": "red", "title": "Compliance Risk Elevated", "desc": "Recent HIPAA audit flag detected in documentation review. Immediate legal review recommended."},
                {"level": "amber", "title": "Revenue Cycle Delay", "desc": "Average claims processing time at 34 days — 9 days above benchmark. Cash flow impact: -$47K/month."},
                {"level": "green", "title": "Patient LTV Strong", "desc": "Patient lifetime value at $3,200 — 18% above regional average. Retention programs showing ROI."},
            ],
            "Retail / E-Commerce": [
                {"level": "red", "title": "Inventory Carrying Cost", "desc": "Excess inventory detected in Q3 planning. Liquidation strategy needed to preserve margin."},
                {"level": "amber", "title": "ROAS Compression", "desc": "Paid channel ROAS dropped to 2.8x from 4.1x. Algorithmic changes require bid strategy revision."},
                {"level": "amber", "title": "Cash Burn in Q3", "desc": "Seasonal dip projected to compress cash position by 18%. Credit facility activation recommended."},
                {"level": "green", "title": "AOV Growing", "desc": "Average order value up 12% MoM — bundling strategy working. Continue and expand."},
            ],
        }

        risk_alerts = risk_templates.get(industry, [
            {"level": "amber", "title": "Revenue Concentration Risk", "desc": "Top 3 clients represent 62% of revenue. Diversification strategy needed."},
            {"level": "amber", "title": "Operational Cost Pressure", "desc": "OpEx growing 3x faster than revenue. Process automation audit recommended."},
            {"level": "green", "title": "Gross Margin Stable", "desc": "Margins holding at 58% — within target range. Monitor for Q4 expansion."},
        ])

        # Financial projections
        monthly_revenue = revenue / 12
        cash_flow = [monthly_revenue * (1 + random.uniform(-0.05, 0.15)) for _ in range(12)]

        customer_health = random.randint(70, 90)
        market_readiness = random.randint(65, 88)

        unit_economics = {
            "CAC": f"${random.randint(180, 420):,}",
            "LTV": f"${random.randint(1800, 5400):,}",
            "LTV:CAC Ratio": f"{random.uniform(3.2, 7.1):.1f}x",
            "Payback Period": f"{random.randint(7, 14)} months",
            "Gross Margin": f"{random.randint(55, 78)}%",
            "Burn Multiple": f"{random.uniform(0.8, 2.4):.1f}x",
        }

        return {
            "risk_alerts": risk_alerts,
            "cash_flow": cash_flow,
            "customer_health": customer_health,
            "market_readiness": market_readiness,
            "unit_economics": unit_economics,
            "timestamp": datetime.now().strftime("%H:%M:%S"),
        }


# ─────────────────────────────────────────────
# ORCHESTRATOR
# ─────────────────────────────────────────────

class AgentOrchestrator:
    """Coordinates the sequential execution of all 5 AI agents."""

    def __init__(self, business_data: dict):
        self.business_data = business_data
        self.outputs = {}
        self.log = []

    def _log(self, agent: str, message: str, color: str = "#3b82f6"):
        self.log.append({
            "time": datetime.now().strftime("%H:%M:%S"),
            "agent": agent,
            "message": message,
            "color": color,
        })

    def run_all(self, fast: bool = False):
        bd = self.business_data

        # Step 1 — CEO
        self._log("CEO Agent", f"Initializing intelligence sweep for {bd.get('company_name', 'company')}...", "#3b82f6")
        if not fast:
            time.sleep(0.2)
        self._log("CEO Agent", "Parsing business documentation and extracting strategic signals...", "#3b82f6")
        ceo_out = CEO_Agent.run(bd)
        self.outputs["ceo"] = ceo_out
        self._log("CEO Agent", f"✅ Executive mandate set: {ceo_out['mandate'][:60]}...", "#10b981")

        # Step 2 — Strategy
        self._log("Strategy Agent", "Receiving mandate from CEO Agent. Initiating market analysis...", "#8b5cf6")
        self._log("Strategy Agent", f"Querying {bd.get('industry','industry')} vertical intelligence database...", "#8b5cf6")
        strategy_out = Strategy_Agent.run(bd, ceo_out)
        self.outputs["strategy"] = strategy_out
        self._log("Strategy Agent", f"✅ Growth strategy compiled: {strategy_out['strategy']['primary'][:55]}...", "#10b981")

        # Step 3 — Marketing
        self._log("Marketing Agent", "Strategy blueprint received. Generating multi-channel campaigns...", "#06b6d4")
        self._log("Marketing Agent", "Synthesizing ad copy, email funnels, and content calendar...", "#06b6d4")
        marketing_out = Marketing_Agent.run(bd, strategy_out)
        self.outputs["marketing"] = marketing_out
        self._log("Marketing Agent", "✅ Campaign suite ready. 4-week content calendar generated.", "#10b981")

        # Step 4 — Sales
        self._log("Sales Agent", "Marketing assets received. Building outbound sales engine...", "#10b981")
        self._log("Sales Agent", "Constructing pipeline stages, lead scoring, and discovery scripts...", "#10b981")
        sales_out = Sales_Agent.run(bd, marketing_out)
        self.outputs["sales"] = sales_out
        self._log("Sales Agent", f"✅ Revenue opportunity identified: ${sales_out['revenue_opportunity']:,.0f}", "#10b981")

        # Step 5 — Finance
        self._log("Finance Agent", "Receiving all outputs. Running financial risk assessment...", "#f59e0b")
        self._log("Finance Agent", "Modeling cash flow, unit economics, and risk vectors...", "#f59e0b")
        finance_out = Finance_Agent.run(bd, strategy_out)
        self.outputs["finance"] = finance_out
        self._log("Finance Agent", f"✅ Risk assessment complete. {len(finance_out['risk_alerts'])} signals flagged.", "#10b981")

        # CEO final synthesis
        self._log("CEO Agent", "All agent outputs synthesized. Dashboard updating...", "#3b82f6")
        self._log("CEO Agent", "🚀 Aegis Growth OS is LIVE. All systems operational.", "#10b981")

        return self.outputs, self.log

File: src/engine/agents.ts
/**
 * Aegis Multi-Agent Engine — 5 AI Agents Boardroom
 */

// Secure independent placeholder setup for API keys
export const CEO_AGENT_API_KEY = process.env.NEXT_PUBLIC_CEO_AGENT_API_KEY || "aegis_key_ceo_sec_v2_placeholder";
export const STRATEGY_AGENT_API_KEY = process.env.NEXT_PUBLIC_STRATEGY_AGENT_API_KEY || "aegis_key_strategy_sec_v2_placeholder";
export const MARKETING_AGENT_API_KEY = process.env.NEXT_PUBLIC_MARKETING_AGENT_API_KEY || "aegis_key_marketing_sec_v2_placeholder";
export const SALES_AGENT_API_KEY = process.env.NEXT_PUBLIC_SALES_AGENT_API_KEY || "aegis_key_sales_sec_v2_placeholder";
export const FINANCE_AGENT_API_KEY = process.env.NEXT_PUBLIC_FINANCE_AGENT_API_KEY || "aegis_key_finance_sec_v2_placeholder";

export interface CEOOutput {
  health_score: number;
  growth_score: number;
  mandate: string;
  summary: string;
  timestamp: string;
}

export interface StrategyPillars {
  primary: string;
  pillars: string[];
  kpis: Record<string, string>;
  markets: string[];
  competitive_moat: string;
}

export interface StrategyOutput {
  strategy: StrategyPillars;
  milestone_velocity: string;
  timestamp: string;
}

export interface MarketingOutput {
  campaigns: {
    hero_ad: string;
    email_subject: string;
    email_body: string;
    channels: string[];
    hook: string;
  };
  content_calendar: Array<{
    week: string;
    content: string;
    channel: string;
  }>;
  timestamp: string;
}

export interface SalesOutput {
  pipeline: Array<{
    stage: string;
    conversion: string;
    avg_days: number;
    action: string;
  }>;
  outbound_sequence: Array<{
    day: string;
    touch: string;
    goal: string;
  }>;
  discovery_script: string;
  lead_score: number;
  timestamp: string;
}

export interface FinanceOutput {
  cash_flow: Array<{
    month: string;
    value: number;
  }>;
  risk_alerts: Array<{
    title: string;
    desc: string;
    level: 'red' | 'amber' | 'green';
  }>;
  revenue_opportunity: number;
  feasibility_score: number;
  timestamp: string;
}

export interface AgentOutputs {
  ceo: CEOOutput;
  strategy: StrategyOutput;
  marketing: MarketingOutput;
  sales: SalesOutput;
  finance: FinanceOutput;
}

export interface LogEntry {
  time: string;
  agent: string;
  message: string;
  color: string;
}

export interface BusinessData {
  company_name: string;
  industry: string;
  annual_revenue: number;
  target_audience: string;
  primary_goal: string;
  team_size: string;
  doc_text: string;
  // Goal tracking states
  target_revenue?: number;
  timeframe_days?: number;
  starting_capital?: number;
  simulated_revenue?: number;
}

const rand = (min: number, max: number): number => Math.floor(Math.random() * (max - min + 1)) + min;
const randFloat = (min: number, max: number): number => (Math.random() * (max - min) + min);
const now = (): string => new Date().toLocaleTimeString('en-US', { hour12: false });

function checkLag(businessData: BusinessData): boolean {
  const target = businessData.target_revenue || 10000;
  const timeframe = businessData.timeframe_days || 30;
  const current = businessData.simulated_revenue || 2500;
  
  const elapsed = timeframe - 20; // mid-horizon elapsed days assumed = 10
  const expectedRevenue = (target / timeframe) * elapsed;
  return current < expectedRevenue;
}

// ─────────────────────────────────────────────
// 1. CEO AGENT
// ─────────────────────────────────────────────
export const CEO_Agent = {
  name: 'CEO Agent',
  icon: '👔',
  color: '#10b981',
  role: 'Chief Executive Officer',
  responsibility: 'Sets strategic mandates & business goal calibrations.',

  run(businessData: BusinessData, customPrompt?: string): CEOOutput {
    const target = businessData.target_revenue || 10000;
    const timeframe = businessData.timeframe_days || 30;
    const isLagging = checkLag(businessData);

    let mandate = `Directing strategy & implementation units to coordinate and achieve the milestone target goal of $${target.toLocaleString()} over a timeframe of ${timeframe} days.`;
    if (isLagging) {
      mandate = `🔴 Lagging Behind Goal - Strategy Pivot Needed: Ordering immediate pricing bundle adjustments and flash marketing campaigns to recover trajectory!`;
    }

    return {
      health_score: isLagging ? rand(68, 76) : rand(84, 96),
      growth_score: isLagging ? rand(62, 70) : rand(82, 95),
      mandate,
      summary: `CEO mandate initiated to establish and scale ${businessData.company_name} target metrics in the ${businessData.industry} sector.`,
      timestamp: now(),
    };
  },
};

// ─────────────────────────────────────────────
// 2. STRATEGY AGENT
// ─────────────────────────────────────────────
export const Strategy_Agent = {
  name: 'Strategy Agent',
  icon: '🧠',
  color: '#3b82f6',
  role: 'Chief Strategy Architect',
  responsibility: 'Calculates milestone velocities and designs growth models.',

  run(businessData: BusinessData, customPrompt?: string): StrategyOutput {
    const isLagging = checkLag(businessData);
    const target = businessData.target_revenue || 10000;
    const timeframe = businessData.timeframe_days || 30;
    const dailyRate = (target / timeframe).toFixed(0);

    const strategy = {
      primary: isLagging
        ? '🔴 Pivot Active: Emergency pricing adjustments and promotional incentives enabled.'
        : 'Enterprise land-and-expand motion with PLG customer loops',
      pillars: ['Usage-based pricing packages', 'Direct executive referral program'],
      kpis: { 'Daily Velocity Target': `$${dailyRate}/day` },
      markets: ['Mid-market TAM', 'Key accounts vertical'],
      competitive_moat: 'High switching costs via deep product integrations',
    };

    return {
      strategy,
      milestone_velocity: `Required rate: $${dailyRate} per day to hit target goal within ${timeframe} days.`,
      timestamp: now(),
    };
  },
};

// ─────────────────────────────────────────────
// 3. MARKETING AGENT
// ─────────────────────────────────────────────
export const Marketing_Agent = {
  name: 'Marketing Agent',
  icon: '📢',
  color: '#eab308',
  role: 'Chief Marketing Officer',
  responsibility: 'Suggests and designs 360 marketing strategies.',

  run(businessData: BusinessData, customPrompt?: string): MarketingOutput {
    const isLagging = checkLag(businessData);
    const company = businessData.company_name || 'Your Company';

    const campaigns = {
      hero_ad: isLagging
        ? `⚠️ FLASH SALE: Discounting all B2B subscription setups by 35% for the next 48 hours to boost pipeline velocity. Claim your slot now!`
        : `Ready to supercharge your business? ${company} gives you the ultimate AI advantage to scale fast.`,
      email_subject: isLagging ? '⚠️ Trajectory override: 35% flash coupon active' : 'The AI operating system for growth',
      email_body: `Hello,\n\nWe are launching a special cohort for B2B clients.\n\nLet\'s talk demo next week.`,
      channels: ['LinkedIn Ads', 'Google Search', 'WhatsApp Blast'],
      hook: isLagging ? '⚠️ Strategy Pivot: Flash promotion activated!' : 'Next-gen tools for business acceleration.',
    };

    const content_calendar = [
      { week: 'Week 1', content: 'Inbound positioning & ad setups', channel: 'LinkedIn' },
      { week: 'Week 2', content: 'Strategic content magnets', channel: 'Email' },
    ];

    return { campaigns, content_calendar, timestamp: now() };
  },
};

// ─────────────────────────────────────────────
// 4. SALES AGENT
// ─────────────────────────────────────────────
export const Sales_Agent = {
  name: 'Sales Agent',
  icon: '🎯',
  color: '#f97316',
  role: 'Chief Revenue Officer',
  responsibility: 'Sets funnel conversion pathways and outreach sequences.',

  run(businessData: BusinessData, customPrompt?: string): SalesOutput {
    const isLagging = checkLag(businessData);
    const company = businessData.company_name || 'Your Company';

    const pipeline = [
      { stage: 'ICP Discovery', conversion: '100%', avg_days: 0, action: 'Lead scoring pipeline' },
      { stage: 'Closed Won', conversion: isLagging ? '45%' : '65%', avg_days: 14, action: 'Agreement executed' },
    ];

    const outbound_sequence = [
      { day: 'Day 1', touch: 'LinkedIn outreach + value summary', goal: 'Initial conversation' },
    ];

    let discovery_script = `OPENING:\n"Hi [Name], this is [Rep] from ${company}. We help companies double conversion rates."`;
    if (isLagging) {
      discovery_script += `\n\n[🔴 Pivot active script override]:\n"Since we are running an end-of-week implementation pilot, we can waive our B2B onboarding fee. Can I secure this pilot slot for you today?"`;
    }

    return {
      pipeline,
      outbound_sequence,
      discovery_script,
      lead_score: isLagging ? rand(60, 80) : rand(75, 96),
      timestamp: now(),
    };
  },
};

// ─────────────────────────────────────────────
// 5. FINANCE AGENT
// ─────────────────────────────────────────────
export const Finance_Agent = {
  name: 'Finance Agent',
  icon: '💵',
  color: '#ef4444',
  role: 'Chief Financial Officer',
  responsibility: 'Audits starting capital, maps cash burn risks, and calculates feasibility.',

  run(businessData: BusinessData, customPrompt?: string): FinanceOutput {
    const target = businessData.target_revenue || 10000;
    const capital = businessData.starting_capital || 5000;
    const current = businessData.simulated_revenue || 2500;
    const isLagging = checkLag(businessData);

    const cash_flow = Array.from({ length: 6 }, (_, i) => ({
      month: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'][i],
      value: Math.round((capital / 6) * (1 + randFloat(-0.05, 0.15))),
    }));

    const risk_alerts = [
      {
        title: isLagging ? '🔴 Cash flow trajectory warning' : '🟢 Budget within parameters',
        desc: isLagging
          ? `Current simulated revenue of $${current.toLocaleString()} lags behind target baseline rate.`
          : `Budget allocation of $${capital.toLocaleString()} is stable.`,
        level: isLagging ? 'red' as const : 'green' as const,
      },
    ];

    return {
      cash_flow,
      risk_alerts,
      revenue_opportunity: Math.round(target * (isLagging ? 0.75 : 1.35)),
      feasibility_score: isLagging ? rand(58, 68) : rand(80, 96),
      timestamp: now(),
    };
  },
};

export const AGENTS_META = [
  CEO_Agent,
  Strategy_Agent,
  Marketing_Agent,
  Sales_Agent,
  Finance_Agent,
];

export function runAgentOrchestrator(businessData: BusinessData, customPrompt?: string): { outputs: AgentOutputs; log: LogEntry[] } {
  const log: LogEntry[] = [];
  const outputs: Partial<AgentOutputs> = {};

  const addLog = (agent: string, message: string, color: string) => {
    log.push({ time: now(), agent, message, color });
  };

  const isLagging = checkLag(businessData);

  if (customPrompt) {
    addLog(
      '🤖 MASTER COPILOT',
      `User request received. Dispatched objective: "${customPrompt}"`,
      '#737373'
    );
  }

  // 1. CEO Agent
  addLog('CEO Agent', 'Meeting opened. Ordering Strategy & Marketing to align on targets.', '#10b981');
  const ceo = CEO_Agent.run(businessData, customPrompt);
  outputs.ceo = ceo;

  // 2. Strategy Agent
  addLog('Strategy Agent', 'Ingesting business industry type to calculate expected daily milestone rate.', '#3b82f6');
  const strategy = Strategy_Agent.run(businessData, customPrompt);
  outputs.strategy = strategy;

  // 3. Marketing Agent
  addLog('Marketing Agent', 'Constructing 360-degree promotional campaign and brand copywriting calendar.', '#eab308');
  const marketing = Marketing_Engine_Proxy.run(businessData, customPrompt);
  outputs.marketing = marketing;

  // 4. Sales Agent
  addLog('Sales Agent', 'Drafting sales scripts, outbound touchpoints, and cold email templates.', '#f97316');
  const sales = Sales_Engine_Proxy.run(businessData, customPrompt);
  outputs.sales = sales;

  // 5. Finance Agent
  addLog('Finance Agent', 'CFO budget audit. Mapping operational cash flows and feasibility metrics.', '#ef4444');
  const finance = Finance_Agent.run(businessData, customPrompt);
  outputs.finance = finance;

  addLog('CEO Agent', 'All 5 AI agent outputs successfully compiled. Boardroom complete.', '#10b981');

  return { outputs: outputs as AgentOutputs, log };
}

// Proxies to adapt types
const Marketing_Engine_Proxy = {
  run(businessData: BusinessData, customPrompt?: string): MarketingOutput {
    return Marketing_Agent.run(businessData, customPrompt);
  }
};

const Sales_Engine_Proxy = {
  run(businessData: BusinessData, customPrompt?: string): SalesOutput {
    return Sales_Agent.run(businessData, customPrompt);
  }
};

export const DEFAULT_BUSINESS: BusinessData = {
  company_name: 'Aura Wellness',
  industry: 'Technology',
  annual_revenue: 1200000,
  target_audience: 'Mid-market B2B SaaS',
  primary_goal: 'Increase Revenue & ARR',
  team_size: '1–10 (Startup)',
  doc_text: `COMPANY OVERVIEW\nWe build workflow automation tools...`,
};
